pkg/clients/ec2/routetable.go

```diff
@@ -67,3 +79,127 @@ func IsAssociationIDNotFoundErr(err error) bool {
 	}
 	return false
 }
+
+// GenerateRTObservation is used to produce v1beta1.RouteTableExternalStatus from
+// ec2.RouteTable.
+func GenerateRTObservation(rt ec2.RouteTable) v1beta1.RouteTableObservation {
+	o := v1beta1.RouteTableObservation{
+		OwnerID:      aws.StringValue(rt.OwnerId),
+		RouteTableID: aws.StringValue(rt.RouteTableId),
+	}
+
+	if len(rt.Routes) > 0 {
+		o.Routes = make([]v1beta1.RouteState, len(rt.Routes))
+		for i, rt := range rt.Routes {
+			o.Routes[i] = v1beta1.RouteState{
+				State: string(rt.State),
+				Route: v1beta1.Route{
+					DestinationCIDRBlock: aws.StringValue(rt.DestinationCidrBlock),
+					GatewayID:            rt.GatewayId,
+				},
+			}
+		}
+	}
+
+	if len(rt.Routes) > 0 {
+		o.Associations = make([]v1beta1.AssociationState, len(rt.Associations))
+		for i, asc := range rt.Associations {
+			o.Associations[i] = v1beta1.AssociationState{
+				Main:          aws.BoolValue(asc.Main),
+				AssociationID: aws.StringValue(asc.RouteTableAssociationId),
+				State:         asc.AssociationState.String(),
+				Association: v1beta1.Association{
+					SubnetID: asc.SubnetId,
+				},
+			}
+		}
+	}
+
+	return o
+}
+
+// LateInitializeRT fills the empty fields in *v1beta1.RouteTableParameters with
+// the values seen in ec2.RouteTable.
+func LateInitializeRT(in *v1beta1.RouteTableParameters, rt *ec2.RouteTable, target v1beta1.RouteTableParameters) { // nolint:gocyclo
+	if rt == nil {
+		return
+	}
+	in.VPCID = awsclients.LateInitializeStringPtr(in.VPCID, rt.VpcId)
+
+	if len(in.Routes) == 0 && len(rt.Routes) != 0 {
+		routes := make([]v1beta1.Route, len(rt.Routes))
+		for i, val := range rt.Routes {
+			routes[i] = v1beta1.Route{
+				DestinationCIDRBlock: aws.StringValue(val.DestinationCidrBlock),
+				GatewayID:            val.GatewayId,
+			}
+			if target.Routes[i].GatewayIDRef != nil {
+				routes[i].GatewayIDRef = target.Routes[i].GatewayIDRef
+			}
+			if target.Routes[i].GatewayIDSelector != nil {
+				routes[i].GatewayIDSelector = target.Routes[i].GatewayIDSelector
+			}
+		}
+		in.Routes = routes
+	}
+
+	if len(in.Associations) == 0 && len(rt.Associations) != 0 {
+		associations := make([]v1beta1.Association, len(rt.Associations))
+		for i, val := range rt.Associations {
+			associations[i] = v1beta1.Association{
+				SubnetID: val.SubnetId,
+			}
+			if target.Associations[i].SubnetIDRef != nil {
+				associations[i].SubnetIDRef = target.Associations[i].SubnetIDRef
+			}
+			if target.Associations[i].SubnetIDSelector != nil {
+				associations[i].SubnetIDSelector = target.Routes[i].GatewayIDSelector
+			}
+		}
+		in.Associations = associations
+	}
+
+	if len(in.Tags) == 0 && len(rt.Tags) != 0 {
+		in.Tags = v1beta1.BuildFromEC2Tags(rt.Tags)
+	}
+}
```

Comment:
Ideally, there should be a field of the struct and we should late-init only the elements that we don't have locally. But that doesn't handle the fact that user might have wanted to delete that entry. Lately, I've been thinking that maybe we'd require user to prefix `-` in the key field value to let us know that this entry should be removed, similar to kubectl unlabel notation.

The late-init only in case the array is empty is the safe solution but there are cases where there is a constant `default` entry even if you didn't create it, so, we'd be missing that. But I think skipping if the spec array is not empty should be fine for this PR

Reply:
FWIW, this function is not used in `Observe`. Only in UpToDate check as far as I can see.